### ISSUE TITLE
Consider supporting COSE/JOSE security

### ISSUE THREAD

**Issue by laurencelundblade** [ietf-rats-wg/draft-ietf-rats-msg-wrap#156]
The amount of text added to CWT that says how COSE is supported is pretty small. The text that adds COSE to CoSWID is larger, but I don't think all the stuff CoSWID did is necessary.

It seems very clear that COSE and JOSE are the way to do attestation security, the same as EAT does. There's probably no controversy over it or much debate.

Since CMW is very clearly an evidence format, it seems kind of important to have security.

Note also recent discussion on TLS and composite attestation on the rats mailing list. It seems like COSE/JOSE-secured evidence is useful and sometimes preferable for composite attestation evidence compared to TLS.

It seems useful for the rats community to know exactly what security should be used to promote interoperability. It's probably not helpful for 1,000 flowers to bloom (lots of attestation security mechanism).

**Comment by nedmsmith** [comment]
> Since CMW is very clearly an evidence format

But not exclusively and Evidence format...

**Comment by thomas-fossati** [comment]
> +1 I want to add normative text about COSE/JOSE use.

WFM too

> I'm also interested in each cmwc_t imposing its own requirements on protected and unprotected headers.

You mean something like:

_"Specific cmwc_t values MAY impose further requirements on the composition of COSE headers."_

?

> As a baseline I'd require the following protected headers
> 
> For COSE, require COSE_Sign1 and headers
> 
> ```
>   &(alg: 1) => int
>   &(content-type: 3) => "application/cmw+cbor"

OK

> ```
>   &(kid: 4) => bstr
> ```

Not sure about this.  See also the discussion at https://github.com/ietf-rats-wg/draft-ietf-rats-corim/issues/363

> For JOSE, require JWS compact serialization and headers
> 
> ```
>   "alg" => text
>   "cty" => "application/cmw+json"
>   "kid" => text
> ```

ditto (kid)

> We can call these application/cmw+cose and application/cmw+jose, both with a cmwc_t parameter. I'm not particularly confident about the `+jose` versus `+jws` versus `-jws+json` naming though. My preference is just `+jose` even if it's not intended for JWE.

Currently, `+jose` is not a registered [SSS](https://www.iana.org/assignments/media-type-structured-suffix/media-type-structured-suffix.xhtml).
We could request it via the CMW document, like we did for +cwt in the [EAT media-types I-D](https://www.ietf.org/archive/id/draft-ietf-rats-eat-media-type-12.html#section-6.1)

**Comment by deeglaze** [comment]
No argument here.

**Comment by thomas-fossati** [comment]
> No argument here, though I’d only want compact for cmw since that ensures signer signer semantics

Starting simple #167. Let's see if this is acceptable to the JOSE crowd.

**Comment by thomas-fossati** [comment]
@laurencelundblade to check if the issue has been addressed by #166 and #167.

If so, please set it to closed.

**Comment by laurencelundblade** [comment]
Hi, trying to get switched back to CMW mode from EAT AUTH48 mode, but not quite there yet.

Why are alg ID and content types made mandatory? COSE, CoSWID, CWT and EAT don't require them. Seems like over specification.  

I believe the reason COSE, CWT,... don't is to save a few bytes as it is possible to know them from some larger context.

 It seems like CWT should align or have a good reason to differ.

**Comment by thomas-fossati** [comment]
> Hi, trying to get switched back to CMW mode from EAT AUTH48 mode, but not quite there yet.

Nearly there! 🥳 
 
> Why are alg ID and content types made mandatory? COSE, CoSWID, CWT and EAT don't require them. Seems like over specification.
> I believe the reason COSE, CWT,... don't is to save a few bytes as it is possible to know them from some larger context.
> 
> It seems like CWT should align or have a good reason to differ.

It makes sense for COSE and CWT, which are very generic security envelopes, to refrain from making strong requirements.

For EAT, it is also sensible to avoid any such requirement, as the profile mechanism allows compressing many protocol parameters, including crypto algorithms, into a single OID or URI.

CoSWID, OTOH, _does_ specify what goes in the protected headers; see [Section 7 of RFC9393](https://www.rfc-editor.org/rfc/rfc9393.html#section-7).

Regarding the choices currently made:

* alg ID seems a no-brainer to me.  It is small and, interop-wise, very beneficial.
* For content type, COSE makes it a SHOULD.  If the argument is saving a few bytes (which makes total sense to me), we can also allow the corresponding C-F (currently called "TBD1" in the IANA table) alongside the string form.

**Comment by laurencelundblade** [comment]
> For EAT, it is also sensible to avoid any such requirement, as the profile mechanism allows compressing many protocol parameters, including crypto algorithms, into a single OID or URI.

CoSWID, OTOH, does specify what goes in the protected headers; see [Section 7 of RFC9393](https://www.rfc-editor.org/rfc/rfc9393.html#section-7).

Regarding the choices currently made:

alg ID seems a no-brainer to me. It is small and, interop-wise, very beneficial.
For content type, COSE makes it a SHOULD. If the argument is saving a few bytes (which makes total sense to me), we can also allow the corresponding C-F (currently called "TBD1" in the IANA table) alongside the string form.

**Comment by laurencelundblade** [comment]
Also, big thanks for adding COSE to CMW! Really appreciate that!

**Comment by thomas-fossati** [comment]
As a "midpoint solution," I propose:

* Relaxing the MUSTs in _"The protected header MUST include the signature algorithm identifier. The protected header MUST include the content type application/cmw+json."_ to SHOULDs.
* Allowing the CoAP C-F alongside the media type: `"cty": "application/cmw+json" / TBD1`

WDYT?

**Comment by deeglaze** [comment]
The signature structure doesn't have enough context in it to distinguish signing different kinds of documents when the content type is dropped. That said, for a specification I guess we want to just rely on COSE usage profiles more generally than have idiosyncratic requirements on a per-content-type basis.

**Comment by thomas-fossati** [comment]
> The signature structure doesn't have enough context in it to distinguish signing different kinds of documents when the content type is dropped. That said, for a specification I guess we want to just rely on COSE usage profiles more generally than have idiosyncratic requirements on a per-content-type basis.

Is there a spec for "COSE usage profiles" that a CMW use could use?

> Edit: regarding the C-F alongside the media type, I don't understand. Do you just mean using an integer where JWS always has a string? I don't know why you'd use JSON for a constrained environment when CBOR is there. We can drop the content-type protected header in COSE when the payload is tagged with the application/cmw+cbor TN.

The suggestion to also allow the C-F version of "application/cmw+cbor" in COSE's `cty` is to save bytes in constrained use cases.

**Comment by thomas-fossati** [comment]
> This begs the question about applying EAT profiles to CMW. I haven't thought that through yet.

For cbor-record/-tag CMWs that encapsulate EATs, the EAT profile applies.  Possibly, manifesting itself via the `eat_profile` parameter in the media type of the CMW.

For cbor-/json-collection CMWs, the `__cmwc_t` can be used as a "super" profile.

**Comment by thomas-fossati** [comment]
@laurencelundblade can you please comment on the proposal https://github.com/ietf-rats-wg/draft-ietf-rats-msg-wrap/issues/156#issuecomment-2646059143

**Comment by thomas-fossati** [comment]
@laurencelundblade: in #186, I have added the CoAP C-F as per https://github.com/ietf-rats-wg/draft-ietf-rats-msg-wrap/issues/156#issuecomment-2646059143.

I am hesitant about relaxing the mandatory headers though, as I don't think it's buying anything in terms of compression - especially if we allow using the C-F codepoint instead of the media type string in the `cty`.

**Comment by nedmsmith** [comment]
> CMW is officially a RATS evidence format

Not sure if its relevant to the mesg-wrap draft, but CMW is any RATS conceptual message type.

**Comment by nedmsmith** [comment]
The related PRs have been merged or I've approved them. So OK to close this issue.

**Comment by thomas-fossati** [comment]
> The related PRs have been merged or I've approved them. So OK to close this issue.

Agree.  @laurencelundblade if you think there is anything left to discuss feel free to open a new issue.
